Add rules for flannel and kube-proxy traffic and re-enable support for firewalld

jlebon opened the pull request:
Small patches to get flannel to work on Atomic.

eparis commented on ansible/roles/flannel/handlers/main.yml:
```diff
@@ -5,7 +5,6 @@
     - stop docker
     - delete docker0
     - start docker
-  when: inventory_hostname in groups['nodes']
```
docker is no necessarily install on non-nodes, so this is not acceptable as I recall...

jlebon replied:
Understood. As it stands, the docker bridge on the master is on the wrong subnet, but I suppose as long as we don't actually use docker there, it can last until the next reboot.

eparis commented on ansible/roles/flannel/tasks/firewalld.yml:
```diff
@@ -0,0 +1,5 @@
+# https://bugzilla.redhat.com/show_bug.cgi?id=1033606 and I think others say firewalld+docker == bad
+# It is also disabled by the node role.
+- name: disable firewalld
+  service: name=firewalld enabled=no state=stopped
```
Is there an actual incompatibility here? Docker has been fixed. If there is a flannl incompatibility here fix it, don't paper over it. Let kube turn off firewalld (as it does today) until https://github.com/kubernetes/kubernetes/pull/12396 lands.

jlebon replied:
So should we reverse the similar disabling done in the nodes role as well?

eparis replied:
or dream up a way to know if docker is running....   @detiber

detiber replied:
Okay, so within openshift-ansible we disabled firewalld across the board, mainly as a way to ensure that it is disabled before we install, start and configure iptables. The tricky part being that iptables doesn't provide a native or supported way to atomically update and persist firewall rules. It looks like the --wait flag was added as a workaround, but it seems like we still have plenty of room for race conditions with updating the firewall rules.

We sort of worked around this by only updating a single chain, to limit the potential for rule change collisions to happen only when adding the jump rule to our new chain.

With that background out of the way...  openshift-sdn currently manages adding it's own port to the firewall (https://github.com/openshift/openshift-sdn/blob/1694e9adbaae43a42944de1c5243a12dca46a535/ovssubnet/controller/kube/bin/openshift-sdn-kube-subnet-setup.sh for example), so we aren't currently managing those rules through openshift-ansible.

As for detecting docker running, I believe there is also code for that in openshift-sdn when setting up the bridge, but @rajatchopra would be able to better speak to that.

I'd say the more you can avoid messing with iptables rules in ansible the better.

jlebon committed:
ansible/roles/flannel: always try to restart docker

We don't want to put a conditional on the flannel task itself.
Otherwise, flannel will only be restarted on the nodes.

Additionally, we don't want to restart docker only on the nodes, but on
every host on which flannel and docker is installed. To do this, we
rely on the fact that if docker is running, then it means that the
docker0 interface will exist. Checking this is easy since the setup
module that is run at startup aggregates the available interfaces in the
ansible_interfaces variable.

jlebon committed:
ansible/roles/flannel: allow flannel traffic through

This patch adds two iptables rules. The first one is:

    -I INPUT 1 -p udp --dport 8472 -j ACCEPT

This rule ensures that the vxlan port is open, which is needed by
flannel when using the vxlan backend. An equivalent rule is added when
firewalld is present.

The second one is:

    -I FORWARD 1 -i flannel.1 -o docker0 -j ACCEPT

This is required so that packets decapsulated by flannel can pass
through to the docker bridge. We use direct rules here when firewalld is
present.

jlebon committed:
ansible/roles/node: re-enable firewalld

Docker and kubernetes should be able to play nice with firewalld now. So
let's stop simply disabling it and instead open up the required ports.

jlebon committed:
ansible/roles/tasks: allow kube-proxy redirected traffic

The kube-proxy adds REDIRECT rules to the PREROUTING chain of the nat
table so that containers are able to access kubernetes services with IPs
in the fictitious service subnet. However, when these rules are
triggered, the resulting packets will be addressed to the docker bridge
and thus would go through the INPUT chain of the filter table, where
they would get blocked.

This patch adds a rule to allow these packets by letting through packets
originating from the docker bridge.

jlebon commented:
I rebased and updated the PR as follow:

1. Detect whether docker was running before restarting it in the flannel role.
2. Re-enable firewalld support in the node role.
3. Implement firewalld rules for the flannel traffic.
4. Add rules to allow kube-proxy redirected traffic coming from the docker containers.

I successfully tested this on the latest AH tree, as well as with a custom tree with firewalld.

eparis commented on ansible/roles/flannel/handlers/main.yml:
```diff
@@ -5,16 +5,17 @@
 
 - name: stop docker
   service: name=docker state=stopped
+  when: '"docker0" in ansible_interfaces'
```
awesome/WAY better way to solve the problem.
